optimum/habana/transformers/models/idefics2/modeling_idefics2.py

```diff
@@ -0,0 +1,340 @@
+            batch_size, seq_length, _ = inputs_embeds.shape
+        else:
+            raise ValueError("You have to specify either input_ids or inputs_embeds")
+
+        past_seen_tokens = 0
+        return_legacy_cache = True
+        use_new_cache = False  # Ignoring new Cache path for HPU
+        if use_cache and use_new_cache:
+            if not isinstance(past_key_values, Cache):  # kept for BC (non `Cache` `past_key_values` inputs)
+                return_legacy_cache = True
+                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+            past_seen_tokens = past_key_values.get_seq_length()
+        else:
+            if past_key_values is not None:
+                past_seen_tokens = past_key_values[0][0].shape[2]
+        if inputs_embeds is not None and input_ids is None and past_seen_tokens == 0:
+            raise ValueError("When first calling the model, if input_embeds are passed, input_ids should not be None.")
+
+        if inputs_embeds is None:
+            inputs_embeds = self.text_model.get_input_embeddings()(input_ids)
+
+        # START VISUAL INPUTS INTEGRATION
+        if pixel_values is not None and image_hidden_states is not None:
+            raise ValueError("You cannot specify both pixel_values and image_hidden_states at the same time")
+        elif pixel_values is not None:
+            batch_size, num_images, num_channels, height, width = pixel_values.shape
+            pixel_values = pixel_values.to(dtype=self.dtype)  # fp16 compatibility
+            pixel_values = pixel_values.view(batch_size * num_images, *pixel_values.shape[2:])
+
+            # Remove padding images - padding images are full 0.
+            nb_values_per_image = pixel_values.shape[1:].numel()
+            real_images_inds = (pixel_values == 0.0).sum(dim=(-1, -2, -3)) != nb_values_per_image
+            pixel_values = pixel_values[real_images_inds].contiguous()
+
+            # Handle the vision attention mask
+            if pixel_attention_mask is None:
+                pixel_attention_mask = torch.ones(
+                    size=(pixel_values.size(0), pixel_values.size(2), pixel_values.size(3)),
+                    dtype=torch.bool,
+                    device=pixel_values.device,
+                )
+            else:
+                # Remove padding images from the mask/pP p
+                pixel_attention_mask = pixel_attention_mask.view(
+                    batch_size * num_images, *pixel_attention_mask.shape[2:]
+                )
+                pixel_attention_mask = pixel_attention_mask[real_images_inds].contiguous()
+
+            patch_size = self.config.vision_config.patch_size
+            patches_subgrid = pixel_attention_mask.cpu().unfold(dimension=1, size=patch_size, step=patch_size)
+            patches_subgrid = patches_subgrid.cpu().unfold(dimension=2, size=patch_size, step=patch_size)
+            patch_attention_mask = (patches_subgrid.sum(dim=(-1, -2)) > 0).bool()
```

Comment:
Hi, can you please let me know the values of patch_size,
and values/shape of pixel_attention_mask
Trying to see if we can rewrite it in some way so that it stays on HPU

Reply:
the patch_size is 14 for idefics2-8b. see https://huggingface.co/HuggingFaceM4/idefics2-8b/blob/main/config.json#L28
for the pixel_attention_mask. it's mask about pixel attention, shape not fixed, shape is (batch*image_num, image_height, image_width], image heigtht/width is rescale to the scale according to longest edge and shorted edge in config see https://huggingface.co/HuggingFaceM4/idefics2-8b/blob/main/preprocessor_config.json#L22-L25 and code https://github.com/huggingface/transformers/blob/main/src/transformers/models/idefics2/image_processing_idefics2.py#L48. image_num is 5 since image spit in on in config, see https://huggingface.co/HuggingFaceM4/idefics2-8b/blob/main/preprocessor_config.json#L3

Reply:
I think we can do this with a convolution, something like the following. Note I just tried a small example with some random numbers, and did not verify it on HPU ir did not verify on the idefics pipeline. Could you check if it provides any perf benefit (and hopefully doesn't cause any accuracy issues)


```

import torch

bs = 2
patch_size = 14
image_num =5
W=128
H=128

pixel_attention_mask = torch.rand([bs*image_num, W, H])

patches_subgrid = pixel_attention_mask.cpu().unfold(dimension=1, size=patch_size, step=patch_size)
patches_subgrid = patches_subgrid.cpu().unfold(dimension=2, size=patch_size, step=patch_size)
res1 = patches_subgrid.sum(dim=(-1, -2))


filter = torch.ones([1, 1, patch_size, patch_size])

res2 = torch.nn.functional.conv2d(pixel_attention_mask.unsqueeze(1), filter, stride=patch_size).squeeze()

print('max abs', torch.max(torch.abs(res1 - res2)))
```

Reply:
res2 = torch.nn.functional.conv2d(pixel_attention_mask.unsqueeze(1), filter, stride=patch_size).squeeze()
NotImplementedError: convolution_overrideable not implemented. You are likely triggering this with tensor backend other than CPU/CUDA/MKLDNN, if this is intended, please use TORCH_LIBRARY_IMPL to override this function

Reply:
so my above impl works on CPU, and I meant it as a reference. didnt really try it in hpu

for hpu i suppose u have to create a conv layer (nn.module) in the init of the model, set it to weights=1 and then use that instead of the functional API for conv

If that seems too much for this PR, we can skip this, and we can try it out later for a separate PR, so I'll approve this one for now. But if u do run the expt mentione dabove (conv nnmodule in the init instead of torch.nn.functional.conv2d) and get some speed up, then please let us know here, and we can then incorporate it into this PR itself.

Reply:
yes, agree. we could skip this now. meanwhile, I add finetune example of image-to-text. which is used for idecfics finetune. please also help review it.
some issue find during enabling.
1) the image size from dataset varies. so the visual model input is varying. so I move the visual modeling handing to prepare input(only impact 1th token generation) 
2) if ignore_eos is False and batch_size>1. the eos is handled incorrectly now, which will cause accuracy in the evaluation low. I fix it in generation code.
3) I also add testcase for finetune case.